builtin/providers/aws/resource_aws_api_gateway_api_key.go

```diff
@@ -0,0 +1,225 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/apigateway"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsApiGatewayApiKey() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsApiGatewayApiKeyCreate,
+		Read:   resourceAwsApiGatewayApiKeyRead,
+		Update: resourceAwsApiGatewayApiKeyUpdate,
+		Delete: resourceAwsApiGatewayApiKeyDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"enabled": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"stage_key": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"api_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"stage_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsApiGatewayStageKeys(d *schema.ResourceData) []*apigateway.StageKey {
+	var stageKeys []*apigateway.StageKey
+
+	if stageKeyData, ok := d.GetOk("stage_key"); ok {
+		params := stageKeyData.(*schema.Set).List()
+		for k := range params {
+			data := params[k].(map[string]interface{})
+			stageKeys = append(stageKeys, &apigateway.StageKey{
+				RestApiId: aws.String(data["api_id"].(string)),
+				StageName: aws.String(data["stage_name"].(string)),
+			})
+		}
+	}
+
+	return stageKeys
+}
+
+func resourceAwsApiGatewayApiKeyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	log.Printf("[DEBUG] Creating API Gateway API Key")
+
+	apiKey, err := conn.CreateApiKey(&apigateway.CreateApiKeyInput{
+		Name:        aws.String(d.Get("name").(string)),
+		Description: aws.String(d.Get("description").(string)),
+		Enabled:     aws.Bool(d.Get("enabled").(bool)),
+		StageKeys:   resourceAwsApiGatewayStageKeys(d),
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating API Gateway: %s", err)
+	}
+
+	d.SetId(*(*apiKey).Id)
+
+	return resourceAwsApiGatewayApiKeyRead(d, meta)
+}
+
+func resourceAwsApiGatewayApiKeyRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).apigateway
+	log.Printf("[DEBUG] Reading API Gateway API Key: %s", d.Id())
+
+	apiKey, err := conn.GetApiKey(&apigateway.GetApiKeyInput{
+		ApiKey: aws.String(d.Id()),
+	})
+	if err != nil {
+		return err
```

Comment:
We need to add guarding against deletion outside of Terraform here, like you did nicely for many other resources.